Create a script to remove leavers

AbraaoMota opened the pull request:
Removing leavers involves removing their section in the
`data/people.yml` as well as clearing out any links to the unboxed
`/people#name-of-person`. This can be fairly longwinded and it's easy to
miss links by hand.
This commit aims to make that easier by adding a rake task to remove a
person from the unboxed site. This is done by running `rake
remove_person`
 from the root directory. You can then type in the full name of the person 
to remove (as written in the `people.yml` - i.e. Abraao Mota, not abraao-mota),
and the script does the rest.

AbraaoMota committed:
Create a script to remove leavers

Removing leavers involves removing their section in the
`data/people.yml` as well as clearing out any links to the unboxed
`/people#name-of-person`. This can be fairly longwinded and it's easy to
miss links by hand.
This commit aims to make that easier by adding a rake task to remove a
person from the unboxed site. This is done by running `rake
remove_person` from the root directory. You can then type in the full
name of the person to remove (as written in the `people.yml` - i.e.
Abraao Mota, not abraao-mota), and the script does the rest.

neilvanbeinum commented on lib/remove_person.rb:
```diff
@@ -0,0 +1,91 @@
+  File.open(people_yml_path, "r") do |input|
+    to_remove_block = false
+    input.each_line do |line|
+      if to_remove_block && line.strip.empty?
```
A bit of a nitpick really but... this works fine for detecting the end of the block for the current state of our `people.yml` where the details of each person in the list happens to be separated by a blank line. But I think it would be a little more robust if we instead check for the next item at the same level in the list (a `-` at the start of the line should do it) to cover the case where someone adds a new person's details without this blank line.

AbraaoMota replied:
@NeilvB  Ah, sure that makes sense. I thought the syntax for separation actually required the blank line, but this sounds more robust.

h-lame replied:
As an alternative, could we treat the file as a datastore and manipulate the data directly, rather than treating it as a syntax file to be parsed.  What I mean is that we use the YAML library to load the contents of people.yml file into a hash, we manipulate the data in the hash, and then use the YAML library to write it back out back into the people.yml file.  That may not preserve the order, but I don't know if the order is important.  If it is important and doing this would change that order then we can make the order more explicit in the data (with an order attribute perhaps?) instead of relying on file order.

neilvanbeinum commented on lib/remove_person.rb:
```diff
@@ -0,0 +1,91 @@
+ag_people_page_link_regex = "'.*\(\D*\/people##{short_name(name)}\).*'"
+dead_link_count = 0
+
+source_files = `ag #{ag_people_page_link_regex} source/ -l`.split("\n")
```
Do we expect all of the unboxed.co devs to have the `ag` command line tool available? Maybe we should add a note that it's a requirement in the readme if not. If there's a more widely-available tool, could we have it do a similar job here? (Not sure of the answers...just pondering!)

AbraaoMota replied:
Good point. I will see if I can make an alternative version, perhaps using `grep` or a derivative of it (that doesn't require prior installation). Otherwise a readme note makes sense

neilvanbeinum commented on lib/remove_person.rb:
```diff
@@ -0,0 +1,91 @@
+        if people_page_link_regex.match(line)
+          partitioned_line = line.partition(/\[#{Regexp.quote first_name(name)}\D*?\]/)
+          remainder_of_line = partitioned_line[2].partition(/\(.*\/people##{Regexp.quote short_name(name)}\)/)
+          clean_line = "#{partitioned_line[0]}#{partitioned_line[1][1..-2]}#{remainder_of_line[2]}"
```
I wonder if we could use `gsub` here. I think it might read a little neater than partitioning the line and then rebuilding. I think we should also be able to handle cases where there's variation in how we refer to people (I noticed we use `[Chris C](<person-link>)` and `[Crystal](<person-link>)` variably, for instance).

I was thinking of something like:

```
profile_link_regex = /\[[^\]]*\]\([^\)]*\/people##{Regexp.quote short_name(name)}\)/
clean_line = line.gsub(profile_link_regex, first_name(name))
```

I'm not a regex expert so there might be a better way. Jack showed me http://rubular.com/ which I used for testing this, in case you haven't come across it.

Test console output in case useful:

```
name = "neil-van-beinum"

line = "[Jack](https://unboxed.co/people/#jack-bracewell) and [Neil](https://unboxed.co/people#neil-van-beinum) used this time to build an app to facilitate our [Thieving Santa](http://www.downhomelife.com/article.php?id=1109) game for the Unboxed Christmas party.<br/>"

line.gsub(/\[[^\]]*\]\([^\)]*\/people##{Regexp.quote short_name(name)}\)/, first_name(name))

"[Jack](https://unboxed.co/people/#jack-bracewell) and Neil used this time to build an app to facilitate our [Thieving Santa](http://www.downhomelife.com/article.php?id=1109) game for the Unboxed Christmas party.<br/>"
```

AbraaoMota replied:
That's way neater yes!

neilvanbeinum commented on README.md:
```diff
@@ -32,6 +32,9 @@ To get things running on your machine:
 * ### Making a news post
     Use `middleman article "[article name]" --blog news` to create a new blog file.
 
+* ### Removing someone from people
```
👍

neilvanbeinum reviewed (changes requested):
This is a great idea and will save us a bunch of time! I think there's a couple of things to check out further:

I noticed that we aren't handling cases where the person can be referred to by various names/nicknames. I suggested a possible way to handle this above.

I wonder if we can also handle cases where the user is referred to in case study pages. E.g. in https://github.com/unboxed/unboxed.co/blob/production/source/project-stories/sh24.erb#L107 there's a link to my profile. This is detected by the regex but re-written to the new file unchanged in the current implementation. Maybe we could add checking for `link: "<blah>"` and avoid writing the line in such cases?

Commit:
Remove rake dependency, and call ruby script directly

ubxd commented:
Deploy preview for *unboxed-co-production* ready!

Built with commit 9026ec7ecb5c4042a3533f9c39f42d256044fe82

https://deploy-preview-311--unboxed-co-production.netlify.com

Commit:
Refactor removal of person from data/people.yml

Simplify parsing YAML file by parsing content as a YAML object in Ruby,
then remove the hash object and rewrite back to the original file

Commit:
Refactor removal of person from posts and pages

Remove dependency on `ag` command and use system's `grep` command to
find files instead. Also target only product stories to remove links to
the page page. Currently, we do not remove any other links (for example,
blog posts).

tomsabin commented on lib/remove_person.rb:
```diff
@@ -0,0 +1,35 @@
+puts "Enter name of person to remove:"
+name = gets.strip!
+short_name = name.gsub(" ", "-").downcase
+first_name = name.split(" ")[0]
```
This variable is unused - removing.

Commit:
Regenerate people.yml from new script that modifies file

When running the `lib/remove_person.rb` script, the people.yml file is
rewritten using the YAML library. Thankfully, order is preserved when
this happens, but we do lose some of the formatting that we've added in
(quotation marks, new lines, etc.) However, should should not be
problematic if we continue to use scripts to modify this file.

neilvanbeinum commented on README.md:
```diff
@@ -50,6 +50,10 @@ Using [BrowserStack](https://www.browserstack.com) for compatibility testing
 
 Upload caricatures at a size at or near to (212px * 425px), with a transparent background. This can be done with e.g. the Mac Preview app.
 
+## Removing someone from people
```
Can we make it something like 'Removing an employee from the site'?

Commit:
Update readme and script instructions
